File: patches/fix_asar_folder_drop.py
#!/usr/bin/env python3
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: python
"""
Prevent app.asar from being dispatched to Cowork on Linux.

When Claude Desktop is re-launched (e.g. by an OAuth callback,
claude:// URL handler, or clicking the launcher while running), the
second instance sends its argv to the running first instance via
Electron's second-instance event. The argv contains the app.asar path.

On Linux, Electron's ASAR-aware filesystem makes .asar files appear as
valid paths via existsSync()/statSync(), so app.asar from process.argv
passes through the existing filters and gets dispatched as a dropped
file on every launch.

The existing __cdb_isAsarPath / __cdb_sanitizeCwd patches catch the CWD
and isDirectory (folder detection) cases, but don't cover the file-drop
path through noe().

Fix (two layers):
1. Filter .asar paths at the top of noe() — this is the single
   convergence point for ALL file-drop code paths (second-instance,
   startup argv, open-file events), so one guard catches everything.
2. Guard the second-instance argv parser (KXn) to skip app.asar early,
   preventing unnecessary processing.

Credit: @dvolonnino for identifying the noe() convergence point.
See: https://github.com/patrickjaja/claude-desktop-bin/issues/24

Usage: python3 fix_asar_folder_drop.py <path_to_index.js>
"""

import sys
import os
import re


def patch_asar_folder_drop(filepath):
    """Filter app.asar paths from file-drop dispatch."""

    print("=== Patch: fix_asar_folder_drop ===")
    print(f"  Target: {filepath}")

    if not os.path.exists(filepath):
        print(f"  [FAIL] File not found: {filepath}")
        return False

    with open(filepath, "rb") as f:
        content = f.read()

    # ── Idempotency check ────────────────────────────────────────────
    # Check for the noe() .asar filter (primary fix)
    if re.search(rb"function [\w$]+\([\w$]+\)\{[\w$]+=[\w$]+\.filter\([\w$]+=>!/\\\.asar/", content):
        print("  [SKIP] Already patched (.asar filter found)")
        return True

    original_content = content

    # ── 1. Patch file-drop convergence point ─────────────────────────
    #
    # All file-drop code paths (second-instance argv, startup argv,
    # open-file events) converge through a single function (noe/Coe/etc,
    # name changes each release). Filter .asar paths here so nothing
    # slips through.
    #
    # Before:
    #   function Coe(t){if(R.info(`Handling file drop: ${t.join(", ")}`),...
    #
    # After:
    #   function Coe(t){t=t.filter(f=>!/\.asar/.test(f));if(!t.length)return;if(R.info(`Handling file drop: ${t.join(", ")}`),...

    pat_noe = (
        rb"(function [\w$]+\()([\w$]+)(\)\{)"
        rb"(if\([\w$]+\.info\(`Handling file drop:)"
    )

    def repl_noe(m):
        arg = m.group(2)
        return m.group(1) + arg + m.group(3) + arg + rb"=" + arg + rb".filter(f=>!/\.asar/.test(f));if(!" + arg + rb".length)return;" + m.group(4)

    content, count = re.subn(pat_noe, repl_noe, content, count=1)

    if count > 0:
        print(f"  [OK] noe() file-drop filter: {count} match(es)")
    else:
        print("  [FAIL] noe() pattern: 0 matches")
        return False

    # ── 2. Guard second-instance argv parser (defense-in-depth) ──────
    #
    # The KXn function (second-instance argv parser) iterates
    # argv.slice(1) and dispatches paths. Unlike the first-instance
    # startup code (VXn) which filters via `path.resolve(s) !== appPath`,
    # KXn has no such filter.
    #
    # Before:
    #   for(const n of t.slice(1))if(!GXn(n)){...
    #
    # After:
    #   for(const n of t.slice(1))if(!/\.asar/.test(n)&&!GXn(n)){...

    pat_argv = (
        rb"(for\(const )([\w$]+)( of [\w$]+\.slice\(1\)\))"
        rb"(if\()(![\w$]+\(\2\))"
    )

    def repl_argv(m):
        var = m.group(2)
        return m.group(1) + var + m.group(3) + m.group(4) + rb"!/\.asar/.test(" + var + rb")&&" + m.group(5)

    content, count = re.subn(pat_argv, repl_argv, content, count=1)

    if count > 0:
        print(f"  [OK] Second-instance argv parser (KXn): {count} match(es)")
    else:
        print("  [WARN] Second-instance argv parser: 0 matches (noe filter is primary)")

    # ── Write back ───────────────────────────────────────────────────
    if content != original_content:
        with open(filepath, "wb") as f:
            f.write(content)
        print("  [PASS] Patch applied")
        return True
    else:
        print("  [WARN] No changes made")
        return True


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <path_to_index.js>")
        sys.exit(1)

    success = patch_asar_folder_drop(sys.argv[1])
    sys.exit(0 if success else 1)

File: patches/fix_asar_folder_drop.nim
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: nim
# Prevent app.asar from being dispatched to Cowork on Linux.
# Two fixes: noe() file-drop filter and second-instance argv parser guard.
# Uses std/nre for backreference support in patterns.

import std/[os, strformat, options]
import std/nre

const EXPECTED_PATCHES = 2

proc apply*(input: string): string =
  result = input
  var patchesApplied = 0

  # Idempotency check
  let idempotencyPat =
    re"function [\w$]+\([\w$]+\)\{(?:var [\w$]+(?:,[\w$]+)*;)?[\w$]+=[\w$]+\.filter\([\w$]+=>!/\\\.asar/"
  if result.find(idempotencyPat).isSome:
    echo "  [SKIP] Already patched (.asar filter found)"
    return result

  # Patch A: noe() file-drop convergence filter
  # v1.19367.0: minifier emits an optional `var r;` (hoisted temp) between the
  # opening brace and the `if(...)`; the filter injection goes after it.
  let patNoe =
    re"(function [\w$]+\()([\w$]+)(\)\{(?:var [\w$]+(?:,[\w$]+)*;)?)(if\([\w$]+\.info\(`Handling file drop:)"

  let mA = result.find(patNoe)
  if mA.isSome:
    let m = mA.get
    let arg = m.captures[1]
    let replacement =
      m.captures[0] & arg & m.captures[2] & arg & "=" & arg &
      ".filter(f=>!/\\.asar/.test(f));if(!" & arg & ".length)return;" & m.captures[3]
    result =
      result[0 ..< m.matchBounds.a] & replacement & result[m.matchBounds.b + 1 .. ^1]
    echo "  [OK] noe() file-drop filter: 1 match(es)"
    patchesApplied += 1
  else:
    echo "  [FAIL] noe() pattern: 0 matches"
    raise newException(ValueError, "fix_asar_folder_drop: noe() pattern not found")

  # Patch B: Second-instance argv parser guard (uses \2 backreference)
  #
  # v1.9659.4: for(const X of Y.slice(1))if(!Z(X)){...
  # v1.11187.4: the .slice(1).filter(...) was hoisted into a local var, so the
  #   loop is now `for(const X of <var>)if(!Z(X)){if(NiA(X)){hp(X,...,"skill file")`.
  # v1.20186.1: the loop body gained braces and a leading directory-collector
  #   block: `for(const X of Y){if(lM(X)){r??(r=X);continue}if(!Z(X)){if(lF(X)){
  #   ru(X,...,"skill file")`. Fold that new prefix into the pre-guard capture.
  # Anchor on the trailing `"skill file"` arg to keep the match unique (there is
  # exactly one such loop in the bundle).
  let patArgv =
    re"(for\(const )([\w$]+)( of [\w$]+\)\{if\([\w$]+\(\2\)\)\{[\w$]+\?\?\([\w$]+=\2\);continue\})(if\()(![\w$]+\(\2\))(\)\{if\([\w$]+\(\2\)\)\{[\w$]+\(\2,[\w$]+,""skill file""\))"

  let mB = result.find(patArgv)
  if mB.isSome:
    let m = mB.get
    let varName = m.captures[1]
    let replacement =
      m.captures[0] & varName & m.captures[2] & m.captures[3] & "!/\\.asar/.test(" &
      varName & ")&&" & m.captures[4] & m.captures[5]
    result =
      result[0 ..< m.matchBounds.a] & replacement & result[m.matchBounds.b + 1 .. ^1]
    echo "  [OK] Second-instance argv parser: 1 match(es)"
    patchesApplied += 1
  else:
    echo "  [FAIL] Second-instance argv parser: 0 matches"

  if patchesApplied < EXPECTED_PATCHES:
    raise newException(
      ValueError,
      &"fix_asar_folder_drop: Only {patchesApplied}/{EXPECTED_PATCHES} patches applied",
    )

when isMainModule:
  if paramCount() != 1:
    echo "Usage: fix_asar_folder_drop <file>"
    quit(1)
  let file = paramStr(1)
  echo "=== Patch: fix_asar_folder_drop ==="
  echo &"  Target: {file}"
  if not fileExists(file):
    echo &"  [FAIL] File not found: {file}"
    quit(1)
  let input = readFile(file)
  let output = apply(input)
  if output != input:
    writeFile(file, output)
    echo &"  [PASS] Patch applied ({EXPECTED_PATCHES}/{EXPECTED_PATCHES} patches applied)"
  else:
    echo &"  [FAIL] No changes made"
    quit(1)
